### PR TITLE
Implement textNumberPattern 'V' virtual decimal point.

### PR DESCRIPTION
**Pull request by mbeckerle**
Implements 'V' a major missing feature for Cobol data.

I have also updated the DFDLSchemas Cobol example to use a snapshot of this. And that works.
See https://github.com/DFDLSchemas/Cobol/pull/2

DAFFODIL-853

### CODE REVIEW

**Comment by mbeckerle** [comment]
This has changed only for textNumberRep 'standard' now, I want to debug that first, then apply changes to 'zoned' and refactor common stuff.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 1]
```diff
@@ -63,36 +142,72 @@ case class ConvertTextNumberPrim(e: ElementBase)
...
-  val textNumberFormatEv: TextNumberFormatEv = {
-    val (pattern, patternStripped) = {
-      val p = e.textNumberPattern
-
-      if (p.startsWith(";")) {
-        e.SDE("Invalid textNumberPattern: The postive number pattern is mandatory")
-      }
+  private val pattern = {
+    val p = e.textNumberPattern
+    if (p.startsWith(";")) {
```
Comment (mbeckerle) [inline daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:147]: e.SDEUnless would be cleaner

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 2]
```diff
@@ -63,36 +142,72 @@ case class ConvertTextNumberPrim(e: ElementBase)
...
+  private val runtimePattern =
+    if (hasV || hasP) patternWithoutPV else pattern
+
+  val textNumberFormatEv: TextNumberFormatEv = {
 
       // Load the pattern to make sure it is valid
       try {
-        new DecimalFormat(p)
+        if (!hasV && !hasP)
+          new DecimalFormat(pattern) // But only if it doesn't have P nor V in it, since ICU doesn't support those
```
Comment (mbeckerle) [inline daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:198]: Add comment about this check being unnecessary for P, V case since regex checking of pattern allows only a tiny subset of all that ICU's patterns can do, and we're certain this will be error free.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 3]
```diff
@@ -63,36 +142,72 @@ case class ConvertTextNumberPrim(e: ElementBase)
...
-      (p, patternNoQuoted)
-    }
+      // At this point we have 2 patterns
+      // 1. The original textNumberPattern string from the schema
+      // 2. The original but with the P or V removed (since they dont correspond to anything in the data)
+      //
+      // We need to use the original textNumberPattern in diagnostic messages
+      // since that's what the schema author wrote.
+      // But if the pattern has P or V, then we need ICU at runtime to use the original
+      // but with the P or V removed since the P and V don't actually represent anything in the data.
```
Comment (mbeckerle) [inline daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:210]: Move this comment up to the val runtimePattern

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala [hunk 4]
```diff
@@ -99,11 +102,30 @@ case class ConvertTextNumberParser(
 
         // Verify that what was parsed was what was passed exactly in byte count.
         // Use pos to verify all characters consumed & check for errors!
-        if (num == null) {
+        if (num1 == null) {
```
Comment (mbeckerle) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala:105]: Move this null check above the icuNum match.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 5]
```diff
@@ -46,6 +46,85 @@ case class ConvertTextCombinator(e: ElementBase, value: Gram, converter: Gram)
...
+    //  DFDL v1.0 Spec says
+    //  It is a Schema Definition Error if any symbols other than "0", "1" through "9" or #
+    //  are used in the vpinteger region of the pattern.
+    //
+    // The prefix and suffix chars can surround the vpinteger region, and there can be
+    // a positive and a negative pattern.
+    //
+    // The prefix and suffix cannot be digits, # or P or V
+    //
+    val prefix = """([^0-9#PV]?)"""
```
Comment (mbeckerle) [inline daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:65]: Can be stronger. prefix and suffix can only contain anything other than the special characters of the pattern language.

**Comment by mbeckerle** [comment]
Ready for review now. 
Works for textNumberRep standard and zoned. Has tests for both.

File: daffodil-lib/src/main/scala/org/apache/daffodil/util/DecimalUtils.scala [hunk 6]
```diff
@@ -479,7 +479,10 @@ object DecimalUtils {
     decodedValue
   }
 
-  def zonedFromNumber(num: String, zonedStyle: TextZonedSignStyle, opl: OverpunchLocation.Value): String = {
+  def zonedFromNumber(
```
Comment (mbeckerle) [inline daffodil-lib/src/main/scala/org/apache/daffodil/util/DecimalUtils.scala:482]: Revert this change.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala [hunk 7]
```diff
@@ -54,11 +54,54 @@ case class ConvertTextCombinatorParser(
...
+
+  final protected val divisor = scala.math.pow(10.0, textDecimalVirtualPoint)
+  final protected val multiplier = scala.math.pow(10.0, textDecimalVirtualPoint)
+
+  final protected def applyTextDecimalVirtualPointForParse(num1: Number): Number = {
+    if (textDecimalVirtualPoint == 0) num1
+    else {
+      // scale for virtual decimal point
+      val scaledNum: Number = num1 match {
+        case l: java.lang.Long => java.math.BigDecimal.valueOf(l).scaleByPowerOfTen(-textDecimalVirtualPoint)
```
Comment (mbeckerle) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala:68]: Note asymmetry around type Long with unparser.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala [hunk 8]
```diff
@@ -78,18 +121,18 @@ case class ConvertTextNumberParser(
     // will match either all or none of 'str', never part of it. Thus,
     // findFirstIn() either matches and it's a zero rep, or it doesn't and it's
     // not a zero
-    val numValue = zeroRepsRegex.find { _.findFirstIn(str).isDefined } match {
+    val numValue: DataValueNumber = zeroRepsRegex.find { _.findFirstIn(str).isDefined } match {
```
Comment (mbeckerle) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala:124]: Test interactions with zero reps, and for 'standard' with float/double with NaN and infinity reps.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 9]
```diff
@@ -46,8 +46,231 @@ case class ConvertTextCombinator(e: ElementBase, value: Gram, converter: Gram)
...
+  }
+
+  def textDecimalVirtualPointForStandard(
+    e: ImplementsThrowsOrSavesSDE,
+    originalPattern: String,
+    patternStripped: String): Int = {
+    val r = TextNumberPatternUtils.vregexStandard
+    r.findFirstMatchIn(patternStripped) match {
+      case Some(r(_, _, afterV, _, _, _, _, _)) => afterV.length
+      case None =>
```
Comment (mbeckerle) [inline daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:85]: Maybe provide better breakdown of syntax to provide better diagnostics here. 
This may help consolidate all the ad-hoc checking.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 10]
```diff
@@ -46,8 +46,231 @@ case class ConvertTextCombinator(e: ElementBase, value: Gram, converter: Gram)
...
+   *
+   * @param pattern the dfdl:textNumberPattern pattern
+   * @return -1 if the pattern is illegal syntax for use with V (virtual decimal point)
+   *         otherwise the number of digits to the right of the V character.
+   */
+  private[primitives] def textDecimalVirtualPointForZoned(pattern: String): Option[Int] = {
+    val r = TextNumberPatternUtils.vregexZoned
+    r.findFirstMatchIn(pattern) match {
+      // note: cannot have both a prefix and suffix. Only one of them.
+      case Some(r(pre, _, afterV, suf)) if (pre.length + suf.length <= 1) => Some(afterV.length)
```
Comment (mbeckerle) [inline daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:123]: Better diagnostic behavior here.

**Comment by mbeckerle** [comment]
One test failing zoned_textNumberPattern_fail05
due to change in diagnostic behavior. There is a check for "E" in the pattern, but that is being suppressed by some other error detection.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala [hunk 11]
```diff
@@ -78,18 +121,18 @@ case class ConvertTextNumberParser(
...
       case Some(_) => primNumeric.fromNumber(0)
```
Comment (mbeckerle) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala:125]: Also, test with smaller integer types.  Virtual decimal points should work for type xs:byte even.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala (hunk 5 above)
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:67]: Not correct. When 'V' or 'P' are present, these ARE the only allowed special pattern characters.

**Comment by tuxji** [comment]
DAFFODIL-853 says both V and P symbols need support, but this pull request adds support for only V.  Are you planning to make DAFFODIL-853 just about V and create a new issue for P, or what?

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala [hunk 12]
```diff
@@ -54,11 +56,58 @@ case class ConvertTextCombinatorParser(
   }
 }
 
+trait TextDecimalVirtualPointMixin {
+  def textDecimalVirtualPoint: Int
+
+  final protected val divisor = scala.math.pow(10.0, textDecimalVirtualPoint)
+  final protected val multiplier = scala.math.pow(10.0, textDecimalVirtualPoint)
```
Comment (tuxji) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala:63]: I'm confused, how come you are defining both divisor and multiplier as the same value?  Please add a comment explaining why.

**Review by tuxji — APPROVED** [review]
+1

I asked some questions and suggested a few minor changes, but the PR looks good overall.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/ElementBaseGrammarMixin.scala [hunk 13]
```diff
@@ -642,7 +642,7 @@ trait ElementBaseGrammarMixin
 
   private lazy val textConverter = {
     primType match {
-      case _: NodeInfo.Numeric.Kind => ConvertTextNumberPrim(this)
+      case _: NodeInfo.Numeric.Kind => ConvertTextNumberStandardPrim(this)
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/ElementBaseGrammarMixin.scala:645]: Yes, there are two kinds of textNumberRep, standard and zoned. 
The renaming is because this class is specifically about standard numbers, and the corresponding classes about zoned have "Zoned" in their names.

[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/ElementBaseGrammarMixin.scala:645]: I take your point that it should be ConvertTextStandardNumberPrim.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 14]
```diff
@@ -46,8 +46,226 @@ case class ConvertTextCombinator(e: ElementBase, value: Gram, converter: Gram)
...
+    patternStripped: String): Int = {
+    val r = TextNumberPatternUtils.vregexStandard
+    r.findFirstMatchIn(patternStripped) match {
+      case Some(r(_, _, afterV, _, _, _, _, _)) => afterV.length
+      case None =>
+        e.SDE(
+          s"""The dfdl:textNumberPattern '%s' contains 'V' (virtual decimal point).
+             | Other than the sign indicators, it can contain only
+             | '#', then digits 0-9 then 'V' then digits 0-9.
+             | The positive part of the dfdl:textNumberPattern is mandatory.""".stripMargin('|'),
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:90]: That is a good style suggestion for regexs with capture groups generally. 

The other groups are there in case we need/want to create finer grain diagnostics that point at the smaller parts of the overall pattern, but really I needed them to test that it was working at all. Regexs are notoriously difficult.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 15]
```diff
@@ -46,8 +46,226 @@ case class ConvertTextCombinator(e: ElementBase, value: Gram, converter: Gram)
...
+    // Note: for zoned, can only have a positive pattern
+    // Prefix or suffix can only be '+' character, to
+    // indicate leading or trailing sign, and can only
+    // one of those.
+    //
+    // Also we're not allowing the # character, since I think that
+    // makes no sense for zoned with virtual decimal point.
+    //
+    val vPattern = """^(\+?)([0-9]+)V([0-9]+)(\+?)$""" // only positive pattern allowed, only + for prefix/suffix
+    val re = vPattern.r
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:105]: Good suggestion.

File: daffodil-lib/src/main/scala/org/apache/daffodil/util/Numbers.scala [hunk 16]
```diff
@@ -311,4 +311,28 @@ object Numbers {
...
     n.asInstanceOf[AnyRef]
   }
+
+  /**
+   * This is the official way to tell if a JBigDecimal has an integer value.
+   * @param bd A Java BigDecimal
+   * @return true if the value is an integer (has fraction part == 0).
+   */
+  def isIntegerValue (bd: JBigDecimal) = {
+     bd.signum() == 0 || bd.scale() <= 0 || bd.stripTrailingZeros().scale() <= 0;
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-lib/src/main/scala/org/apache/daffodil/util/Numbers.scala:321]: One of the things learned in doing this V feature is that java boxed numbers are awful. They have almost nothing in common, and almost nothing works polymorphically across them. 
This was the best thing I could find online for determining if a decimal value is actually an integer, and this is awful performance wise since stripTrailingZeros creates another BigInteger object just so you can ask about its scale. 

There *should* be a better way. But finding it is not easy.  

I will convert this into a comment discussing this issue that the Java numbers have almost no polymorphic operations. This was all in service of implementing the isZero() method.

File: daffodil-lib/src/main/scala/org/apache/daffodil/util/Numbers.scala [hunk 17]
```diff
@@ -311,4 +311,28 @@ object Numbers {
...
+    n1 match {
+      case d: JDouble => d.doubleValue() == 0.0
+      case f: JFloat => f.floatValue() == 0.0
+      case bd: JBigDecimal => bd.signum() == 0
+      case bi: JBigInt => bi.signum() == 0
+      case l: JLong => l.longValue == 0
+      case i: JInt => i.intValue == 0
+      case s: JShort => s.shortValue == 0
+      case b: JByte => b.byteValue == 0
+    }
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-lib/src/main/scala/org/apache/daffodil/util/Numbers.scala:335]: I will add a default case, as this is supposed to handle the java numbers that we use in our internal infoset representation, which does not use any of the ones you mention here. 

It was definitely a mistake for Scala to not define its own Number base class/trait. One more compromise to reuse of the Java classes.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala (hunk 12 above)
Comment (mbeckerle) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala:63]: Ha! it's really the "magic number" and one code path multiplies by it. the other divides by it. I'll explain.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 18]
```diff
@@ -46,8 +46,226 @@ case class ConvertTextCombinator(e: ElementBase, value: Gram, converter: Gram)
...
+          // That library has only seen the pattern with the P and V
+          // removed. The messages might show that pattern which would
+          // confuse the schema author since that's not the pattern
+          // they wrote.
+          //
+          // So we try to explain....
+          e.SDE(
+            """Invalid textNumberPattern.
+              | The errors are about the pattern with the P (decimal scaling position)
+              | and V (virtual decimal point) characters removed: %s""".stripMargin, ex)
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:258]: I am actually not sure it's possible to exercise this path. The regexs for the V pattern (and P eventually) constrain things pretty strongly in terms of the pattern that's left when it is handed to ICU. 

So this is more of a "just in case" there is something ICU rejects even after we've pre-conditioned it by removing the P and V.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala [hunk 19]
```diff
@@ -54,11 +56,63 @@ case class ConvertTextCombinatorParser(
...
+        if (textDecimalVirtualPoint == 0) n
+        else badType(n)
+      // $COVERAGE-OFF$
+      case _ => Assert.invariantFailed("Not a JNumber")
+      // $COVERAGE-ON$
+    }
+  }
+}
+
 case class ConvertTextNumberParser(
```
Comment (mbeckerle) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PrimitivesTextNumber1.scala:109]: Rename to ConvertTextStandardNumberParser for consistency. The file should also be renamed.

File: daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ConvertTextNumberUnparser.scala [hunk 20]
```diff
@@ -43,16 +47,18 @@ case class ConvertTextCombinatorUnparser(
 case class ConvertTextNumberUnparser(
```
Comment (mbeckerle) [inline daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ConvertTextNumberUnparser.scala:47]: rename to ConvertTextStandardNumberUnparser. Rename file to correspond.

**Review by jadams-tresys — APPROVED** [review]
+1

Looks good!

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/ConvertTextStandardNumberParser.scala [hunk 21]
```diff
@@ -54,11 +56,64 @@ case class ConvertTextCombinatorParser(
...
+      case bd: JBigDecimal => bd.scaleByPowerOfTen(textDecimalVirtualPoint)
+
+      case n: JNumber =>
+        if (textDecimalVirtualPoint == 0) n
+        // $COVERAGE-OFF$ // both badType and the next case are coverage-off
+        else badType(n)
+      case _ => Assert.invariantFailed("Not a JNumber")
+      // $COVERAGE-ON$
+    }
+  }
```
Comment (stevedlawrence) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/ConvertTextStandardNumberParser.scala:107]: Does this function need to clamp the result to an integer? Or does ICU handle that when we pass the result of this function to `df.format(...)`?

**Review by stevedlawrence — COMMENTED** [review]
+1

This has already been merge, but I wanted to review it any way. Suggest any changes are included in changes needed when `P` support is added.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 22]
```diff
@@ -46,8 +45,249 @@ case class ConvertTextCombinator(e: ElementBase, value: Gram, converter: Gram)
   override lazy val unparser = new ConvertTextCombinatorUnparser(e.termRuntimeData, value.unparser, converter.unparser)
 }
 
-case class ConvertTextNumberPrim(e: ElementBase)
-  extends Terminal(e, true) {
+// This is a separate object for unit testing purposes.
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:48]: I should have made this scaladoc.
I should have said "This object exists to expose some code for convenient unit testing that would otherwise be inaccessible."  It's not that the whole object can move into src/test. It's just there so we can do a bunch of unit tests on the regexs it contains.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 23]
```diff
@@ -46,8 +45,249 @@ case class ConvertTextCombinator(e: ElementBase, value: Gram, converter: Gram)
...
+private[primitives]
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:49]: I suggest using it when you would make something private, but want a unit test to be able to test it. 
That's in fact the only thing I've ever seen package private used for.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 24]
```diff
@@ -46,8 +45,249 @@ case class ConvertTextCombinator(e: ElementBase, value: Gram, converter: Gram)
...
+    val vPattern = s"""^${prefix}${beforeV}V${afterV}${suffix}$$""" // only positive pattern allowed
+    val re = vPattern.r
+    re
+  }
+
+  /**
+   * Checks a pattern for suitability with V (virtual decimal point) in the pattern
+   * in the context of zoned textNumberRep.
+   *
+   * This is broken out separately for unit testing purposes.
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:117]: If this was a method of the class it naturally falls in (responsibility-wise), then it would not be visible for convenient unit testing without having to construct a bunch of schema-compiler infrastructure objects or some complex test rig.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 25]
```diff
@@ -46,8 +45,249 @@ case class ConvertTextCombinator(e: ElementBase, value: Gram, converter: Gram)
...
+
+  /**
+   * The apos character (') is the quoting character in ICU patterns (and DFDL textNumberPattern).
+   * This removes any unquoted P or V characters (which are not implemented by ICU)
+   * leaving a pattern string that is suitable for use with ICU.
+   * @param pattern the textNumberPattern string
+   * @return the pattern string with all unquoted P and unquoted V removed.
+   */
+  def removeUnquotedPV(pattern: String) : String = {
+    val uniqueQQString = "alsdflslkjskjslkkjlkkjfppooiipsldsflj"
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:140]: I suspect a split-based algorithm per your suggestion below, is better.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 26]
```diff
@@ -46,8 +45,249 @@ case class ConvertTextCombinator(e: ElementBase, value: Gram, converter: Gram)
...
+    //
+    // I did some tests, and the formulations you think might work
+    // such as from stack-overflow, don't work.
+    // (See test howNotToUseRegexLookBehindWithReplaceAll)
+    //
+    // So we use this brute force technique of replacing all ''
+    // first, so we have only single quotes to deal with.
+    pattern.replaceAll("''", uniqueQQString).
+      replaceAll("(?<!')P", "").
+      replaceAll("(?<!')V", "").
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:157]: This is addressed in https://github.com/apache/daffodil/pull/910

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala (hunk 25 above)
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:140]: Addressed in https://github.com/apache/daffodil/pull/910

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala (hunk 24 above)
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:117]: Addressed in https://github.com/apache/daffodil/pull/910

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala (hunk 22 above)
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:48]: Addressed in https://github.com/apache/daffodil/pull/910

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala [hunk 27]
```diff
@@ -46,8 +45,249 @@ case class ConvertTextCombinator(e: ElementBase, value: Gram, converter: Gram)
...
+  extends Terminal(e, true)
+  with ConvertTextNumberMixin {
+
+  final override protected lazy val textDecimalVirtualPointFromPattern: Int = {
+    val r = TextNumberPatternUtils.vregexStandard
+    r.findFirstMatchIn(patternStripped) match {
+      case Some(r(_, _, afterV, _, _, _, _, _)) => afterV.length
+      case None =>
+        e.SDE(
+          s"""The dfdl:textNumberPattern '%s' contains 'V' (virtual decimal point).
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala:284]: Resolved in https://github.com/apache/daffodil/pull/910

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesZoned.scala [hunk 28]
```diff
@@ -47,71 +45,67 @@ case class ConvertZonedCombinator(e: ElementBase, value: Gram, converter: Gram)
...
-  extends Terminal(e, true) {
-
-  val textNumberFormatEv: TextNumberFormatEv = {
-    val pattern = {
-      val p = e.textNumberPattern
+  extends Terminal(e, true)
+  with ConvertTextNumberMixin {
+
+  final override protected lazy val textDecimalVirtualPointFromPattern: Int = {
+    TextNumberPatternUtils.textDecimalVirtualPointForZoned(patternStripped).getOrElse {
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesZoned.scala:52]: Renamed in https://github.com/apache/daffodil/pull/910

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesZoned.scala [hunk 29]
```diff
@@ -47,71 +45,67 @@ case class ConvertZonedCombinator(e: ElementBase, value: Gram, converter: Gram)
...
+      e.SDE(
+        s"""The dfdl:textNumberPattern '%s' contains 'V' (virtual decimal point).
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesZoned.scala:54]: Resolved in https://github.com/apache/daffodil/pull/910

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/ConvertTextStandardNumberParser.scala [hunk 30]
```diff
@@ -54,11 +56,64 @@ case class ConvertTextCombinatorParser(
   }
 }
 
-case class ConvertTextNumberParser(
+trait TextDecimalVirtualPointMixin {
+  def textDecimalVirtualPoint: Int
+
+  final protected val virtualPointScaleFactor = scala.math.pow(10.0, textDecimalVirtualPoint)
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/ConvertTextStandardNumberParser.scala:62]: resolved in https://github.com/apache/daffodil/pull/910

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/ConvertTextStandardNumberParser.scala (hunk 21 above)
Comment (mbeckerle) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/ConvertTextStandardNumberParser.scala:107]: Changed and comments added about this in https://github.com/apache/daffodil/pull/910.

**Review by mbeckerle — COMMENTED** [review]
@stevedlawrence Your additional review items were addressed in PR
https://github.com/apache/daffodil/pull/910
Which also adds the 'P' character feature (similar to 'V')

**Comment by mbeckerle** [comment]
Actually I blew it creating PR 910.

The PR with the additional fixes per this PR review is 
https://github.com/apache/daffodil/pull/911
